### DISCUSSION TITLE
NSLock Memory leaks

### DISCUSSION THREAD

**Discussion by Slobodianiuk95** [pointfreeco/swift-composable-architecture#3106] [General]
### Description

Hello! I discovered memory leaks using the Leaks tool while developing my own application. At first, I thought it was my mistake, but then I found the same memory leaks in the Todos and VoiceMemos examples. This happens from the start of the application and increases as it is used.

![CleanShot 2024-05-20 at 13 15 47@2x](https://github.com/pointfreeco/swift-composable-architecture/assets/122380009/25245da6-56ad-4a70-9d4b-1fe7e507fb5f)

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

No leaks happend

### Actual behavior

NSLock leaks in Xcode -> Instruments -> Leaks

### Steps to reproduce

1. Open any example app like Todos
2. Install app on device or simulator
3. Open Xcode -> Instruments -> Leaks and start recording Todos app session

### The Composable Architecture version information

1.10.4

### Destination operating system

iOS 17.5

### Xcode version information

Xcode 15.4

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @Slobodianiuk95, this lock is only used in [`#if DEBUG` code](https://github.com/pointfreeco/swift-perception/blob/9f6c437400bead83f98da0d98e129e68914b1e2d/Sources/Perception/PerceptionRegistrar.swift#L19-L22) and so the leak does not happen when building in release mode. 

Since this is not an issue with the library I am going to convert it to a discussion. Feel free to continue the conversation over there!

**Reply by Slobodianiuk95** [reply]
@mbrandonw, thank you for the quick response! Now everything is clear to me
